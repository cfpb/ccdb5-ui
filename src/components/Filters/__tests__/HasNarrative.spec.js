import ReduxHasNarrative, {
  HasNarrative,
  mapDispatchToProps,
} from '../HasNarrative';
import configureMockStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import React from 'react';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';

/**
 *
 * @param initialProps
 */
function setupEnzyme(initialProps = {}) {
  const props = Object.assign(
    {
      toggleFlag: jest.fn(),
      options: {
        isChecked: true,
        phase: '',
      },
    },
    initialProps
  );

  const target = mount(<HasNarrative {...props} />);

  return {
    props,
    target,
  };
}

/**
 *
 * @param query
 */
function setupSnapshot(query = {}) {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    query,
  });

  return renderer.create(
    <Provider store={store}>
      <ReduxHasNarrative />
    </Provider>
  );
}

describe('initial state', () => {
  it('renders without crashing', () => {
    const target = setupSnapshot();
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('pre-check filter based on query', () => {
    const target = setupSnapshot({
      has_narrative: true,
    });
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('checks and disables the filter when searching narratives', () => {
    const target = setupSnapshot({
      searchField: 'complaint_what_happened',
    });
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('component::HasNarrative', () => {
  describe('flag filter changed', () => {
    it('triggers an update when checkbox is clicked', () => {
      const { target, props } = setupEnzyme();
      const input = target.find('#filterHasNarrative');
      input.simulate('change');
      expect(props.toggleFlag).toHaveBeenCalled();
    });
  });

  describe('mapDispatchToProps', () => {
    it('hooks into toggleFlag', () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).toggleFlag();
      expect(dispatch.mock.calls.length).toEqual(1);
    });
  });
});
