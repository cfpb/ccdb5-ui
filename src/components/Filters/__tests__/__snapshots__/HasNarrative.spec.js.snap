// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial state checks and disables the filter when searching narratives 1`] = `
<section
  className="single-checkbox"
>
  <h4>
    Only show complaints with narratives?
  </h4>
  <div
    className="m-form-field m-form-field__checkbox"
  >
    <input
      checked={true}
      className="a-checkbox"
      disabled={true}
      id="filterHasNarrative"
      onChange={[Function]}
      type="checkbox"
      value="has_narrative"
    />
    <label
      className="a-label"
      htmlFor="filterHasNarrative"
    >
      Yes
    </label>
  </div>
</section>
`;

exports[`initial state pre-check filter based on query 1`] = `
<section
  className="single-checkbox"
>
  <h4>
    Only show complaints with narratives?
  </h4>
  <div
    className="m-form-field m-form-field__checkbox"
  >
    <input
      checked={true}
      className="a-checkbox"
      disabled={false}
      id="filterHasNarrative"
      onChange={[Function]}
      type="checkbox"
      value="has_narrative"
    />
    <label
      className="a-label"
      htmlFor="filterHasNarrative"
    >
      Yes
    </label>
  </div>
</section>
`;

exports[`initial state renders without crashing 1`] = `
<section
  className="single-checkbox"
>
  <h4>
    Only show complaints with narratives?
  </h4>
  <div
    className="m-form-field m-form-field__checkbox"
  >
    <input
      checked={false}
      className="a-checkbox"
      disabled={false}
      id="filterHasNarrative"
      onChange={[Function]}
      type="checkbox"
      value="has_narrative"
    />
    <label
      className="a-label"
      htmlFor="filterHasNarrative"
    >
      Yes
    </label>
  </div>
</section>
`;
